tests/e2e/utils/product-editor.js

```diff
@@ -10,6 +10,243 @@ import * as api from './api';
+			return page.locator( '.gla-metabox:visible' ).first();
+		},
+
+		getChannelVisibilityMetaBox() {
+			return page.locator( '#channel_visibility' );
+		},
+
+		getChannelVisibilityHeading() {
+			return this.getChannelVisibilityMetaBox().getByRole( 'heading', {
+				name: 'Channel visibility',
+			} );
+		},
+
+		getProductAttributesHeading() {
+			return this.getPluginPanel().getByRole( 'heading', {
+				name: 'Product attributes',
+			} );
+		},
+
+		getChannelVisibility() {
+			const metaBox = this.getChannelVisibilityMetaBox();
+
+			return {
+				selection: metaBox.getByRole( 'combobox' ),
+				help: metaBox.locator( '.description' ),
+				notice: metaBox.locator( '.sync-status' ),
+				status: metaBox.locator( '.sync-status p' ).nth( 1 ),
+				issues: metaBox.getByRole( 'listitem' ),
+			};
+		},
+
+		getSelectWithTextInput() {
+			const field = page.locator( '.select-with-text-input' ).first();
+
+			return {
+				selection: field.getByRole( 'combobox' ),
+				input: field.getByRole( 'textbox' ),
+			};
+		},
+
+		getDateAndTimeInputs( locator = page ) {
+			const simple = locator.locator(
+				'.gla_attributes_availabilityDate_field'
+			);
+			const variation = locator.locator(
+				'.gla_variation_attributes\\[0\\]_availabilityDate_field'
+			);
+			const field = simple.or( variation );
+
+			return {
+				dateInput: field.locator( 'input[type=date]' ),
+				timeInput: field.locator( 'input[type=time]' ),
+			};
+		},
+
+		getMultipackInput() {
+			return page.locator( '.gla_attributes_multipack_field input' );
+		},
+
+		getAllProductAttributes( locator = page ) {
+			const { dateInput: availabilityDate, timeInput: availabilityTime } =
+				this.getDateAndTimeInputs( locator );
+
+			return {
+				gtin: locator.getByLabel( /\(GTIN\)$/ ),
+				mpn: locator.getByLabel( 'MPN' ),
+				brand: locator.getByLabel( 'Brand', { exact: true } ),
+				condition: locator.getByLabel( 'Condition', { exact: true } ),
+				gender: locator.getByLabel( 'Gender', { exact: true } ),
+				size: locator.getByLabel( 'Size', { exact: true } ),
+				sizeSystem: locator.getByLabel( 'Size system' ),
+				sizeType: locator.getByLabel( 'Size type' ),
+				color: locator.getByLabel( 'Color', { exact: true } ),
+				material: locator.getByLabel( 'Material', { exact: true } ),
+				pattern: locator.getByLabel( 'Pattern', { exact: true } ),
+				ageGroup: locator.getByLabel( 'Age Group', { exact: true } ),
+				multipack: locator.getByLabel( 'Multipack', { exact: true } ),
+				isBundle: locator.getByLabel( 'Is Bundle' ),
+				availabilityDate,
+				availabilityTime,
+				adultContent: locator.getByLabel( 'Adult content' ),
+			};
+		},
+	};
+
+	const asyncActions = {
+		async gotoAddProductPage() {
+			await page.goto( '/wp-admin/post-new.php?post_type=product' );
+			await this.waitForInteractionReady();
+		},
+
+		async gotoEditVariableProductPage() {
+			const variableId = await api.createVariableProduct();
+			await api.createVariationProducts( variableId );
```

Comment:
This can be refactored as well.

Reply:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2363/commits/8a7993308650aae932702368c37b2e9160da2922.